### DISCUSSION TITLE
State of RxJS in 2022

### DISCUSSION THREAD

**Discussion by bmadone** [ReactiveX/rxjs#7043] [General]
Hello guys, I want to discuss RxJS, and it parts in developments right now.

I was an Angular dev for 4+ years, then I switched to React, and in my project we use RxJS a lot for state management, and all business logic. I don't think it's the best idea, because it's super complex, and onboarding for new team members it's a hell.

In Angular world, RxJS is everywhere, but I saw a lot of misusing this concept of reactivity. Because if you want to write clean reactive code everything thing is a steam, and then you combine these steams together, and after all you have 1 subscription that run it, most of the angular dev don't do that, instead that do a lot of subscribe() in their code, that's wrong. So what's yours thoughts about this technology.

In my opinion is good it's useful sometimes, the famous example is autocomplete, but in general it's to complicated to use it properly. What's your thoughts? Do you use it in your companies?

**Comment by cpatti97100** [comment]
I have to say I used it extensively in redux-observables and also as a component state management tool. once I grasped the concepts then it was really pleasant to work with. but yes it required effort.
I need to work with redux-saga now at work, and I find it more confusing, there are lots of operators with different signatures there too https://redux-saga.js.org/docs/api#effect-creators

"new team members" is something that was indeed a problem also for us, but I think if we want to raise the quality bar of our web apps in general, that cannot be a limiting factor any longer. we are hiring software engineers after all, I am sure they can manage given proper guidance

**Comment by voliva** [comment]
I'm on the same boat.

I got first introduced to RxJS when I moved from AngularJS to Angular 2 on the company I was, although I didn't fully understand reactive programming back then. I think there was not really good documentation on how to use RxJS reactively in angular, at least at that time. Essentially the HTTP service returned an Observable instead of a promise, and you could put subjects everywhere to have the UI update when you push new data to them. That was my level of understanding.

Then I moved to another company where I've been using React for the last 6 years now, started with Redux + Redux Observable, and here is where I started learning about reactive programming, but soon enough it felt like the state itself was not reactive. Redux Observable just works as a way of reacting to actions to trigger some side effects, and end up spitting more redux actions, but it was easy to fall in the pitfall of using the redux store as a "subject" and making the actions just a setter of that store (.next on a subject)

Talking this with some of the colleagues we spiked something to manage the state directly with RxJS, and hook it into the react components by tapping on those states. This lets us declare our state reactively and keep it independent from the component tree. It also makes it trivial to code split (something extremely challenging with a redux store).

From that spike we released a library, https://github.com/re-rxjs/react-rxjs, which we keep using and maintaining (adding in more features or making it easier to use). We've been using it in a few projects on our company for the past 2 years, and we also have received some good feedback from others that use it.

I'm on a point where I realised that I don't know how to do any sort of application in vanilla React that's a bit more complex than a todo list. I'm so used of thinking reactively when building state that now it's challenging for me to revert it to imperative way (and even more if you have the common issues of vanilla react in the way: useEffect closures, useMemo hell, having to lift state up, etc.)

**Comment by bmadone** [comment]
I just can't figure out what is the good case for picking RxJS in 2022 for frontend if you don't use Angular. In Angular you don't have much choice, they don't have many choices. But if RxJS will close only in Angular world it's dead end I think.

**Reply by cpatti97100** [reply]
not really, I saw a presentation where the Preact team was thinking about integrating Observables into their library.

here a conversation about reactivity in Preact https://github.com/preactjs/preact/pull/3474

**Reply by josepot** [reply]
@bmarvinb what makes you think that?

Check this out:

- `@angular/core`:  [~3 million weekly downloads](https://www.npmjs.com/package/@angular/core).
- `rxjs`: [over 40 million weekly downloads](https://www.npmjs.com/package/rxjs)

FTR: I'm one of the people that heavily relies on rxjs and I don't do angular.

**Reply by bmadone** [reply]
@josepot I checked a lot of front end jobs, and I found RxJS in description only in angular specific

**Reply by bmadone** [reply]
And again, most of the people use RxJS in a wrong way, with a lot of subscription in imperative style, it is not problem with library itself, but a luck of knowledge in general how to do it correctly

**Reply by bmadone** [reply]
For data management React Query is the best that we have right now, for state management we have a ton of libraries, Jotai for sharing data without context, Zustand modern version of Redux, XState for complex stuff, where is place for RxJS in all of it?

**Reply by voliva** [reply]
In React nothing is set in stone. You use whatever you see more fit based on your project and your team's skills.

Where is place for RxJS in all of it? Well, as I said we use RxJS for everything you mentioned here. Granted it is a minority, probably the most used library for react is still redux toolkit (npm stats show 1.5M/week on react-query, vs 1.9M/week on @reduxjs/toolkit), but that doesn't mean RxJS is a dead end or that it can only be used with Angular.

> But if RxJS will close only in Angular world it's dead end I think.

For many people it's really useful and are still using it. On npm @angular/core shows 3M downloads per week, whereas rxjs has over 40M downloads per week. It's far from being dead.

**Reply by dariomannu** [reply]
@bmarvinb maybe this is exactly the case you were looking for? https://github.com/ReactiveX/rxjs/discussions/7257

I'm just working on my second large and complex project using Rimmel for the components and RxJS for all the state management, all in true functional-reactive style. I think the case is now stronger than ever.
